### PR TITLE
DOCS-357 - Avro and Schema Registry

### PR DESCRIPTION
**Pull request by joel-hamill**
https://confluentinc.atlassian.net/browse/DOCS-357

Staging Server

http://staging-docs.confluent.io.s3-website-us-west-2.amazonaws.com/PR/778/4.1.x/ksql/docs/installation/server-config/avro-schema.html

### CODE REVIEW

File: docs/faq.rst [hunk 1]
```diff
@@ -149,9 +149,9 @@ flag. For more information, see :ref:`common-configs`.
 How do I use Avro data and integrate with Confluent Schema Registry?
 ====================================================================
 
-Configure the ``ksql.schema.registry.url`` to point to Schema Registry (see :ref:`common-configs`).
+Configure the ``ksql.schema.registry.url`` to point to Schema Registry (see :ref:`install_ksql-avro-schema`).
 
-.. important:: To use Avro data with KSQL you must have Schema Registry installed. This is included by default with |cpe|.
+.. important:: To use Avro data with KSQL you must have Schema Registry installed. This is included by default with Confluent Platform Enterprise.
```
Comment (miguno) [inline docs/faq.rst]: Not just with Confluent Platform Enterprise. It's in COS and CE.

File: docs/faq.rst [hunk 2]
```diff
@@ -78,7 +78,7 @@ KSQL currently supports formats:
 
 -  DELIMITED (e.g. comma-separated value)
 -  JSON
--  Avro (requires Confluent Schema Registry and setting ``ksql.schema.registry.url`` in the KSQL configuration file)
+-  Avro (requires Confluent Schema Registry and :ref:`setting ``ksql.schema.registry.url`` in the KSQL configuration file <install_ksql-avro-schema>`)
 
 *Support for Apache Avro is expected soon.*
```
Comment (miguno) [inline docs/faq.rst]: We must remove this ("Avro is expected soon").

File: docs/installation/server-config/avro-schema.rst [hunk 3]
```diff
@@ -1,4 +1,96 @@
...
+    :local:
+
+Supported functionality
+^^^^^^^^^^^^^^^^^^^^^^^
+
+KSQL currently supports Avro data in the Kafka message values.
+
+The following functionality is not yet supported:
+
+-  Message keys in Avro format are not supported. Message keys in KSQL are always interpreted as STRING format, which means
```
Comment (miguno) [inline docs/installation/server-config/avro-schema.rst]: not supported [yet]

File: docs/installation/server-config/avro-schema.rst [hunk 4]
```diff
@@ -1,4 +1,96 @@
...
+   KSQL will ignore Avro schemas that have been registered for message keys.
+-  Avro schemas with nested fields are not supported. This is because KSQL does not yet support nested columns.
```
Comment (miguno) [inline docs/installation/server-config/avro-schema.rst]: "Avro schemas with nested fields are not supported yet. This is because KSQL does not yet support nested columns."

File: docs/installation/server-config/avro-schema.rst [hunk 5]
```diff
@@ -1,4 +1,96 @@
...
+in the KSQL configuration file (``<path-to-confluent>/etc/ksql/ksql-server.properties``) that you use to start the KSQL
+Server. For more information, see :ref:`install_ksql-server`.
+
+.. important:: Do not use the SET statement in the KSQL CLI to configure the registry endpoint.
+
+Using Avro in KSQL
+^^^^^^^^^^^^^^^^^^
+
+Before using Avro in KSQL, make sure that |sr| is up and running and that ``ksql.schema.registry.url`` is set correctly
+in the KSQL properties file. |sr| is :ref:`installed by default <quickstart>` with |cpe|.
```
Comment (miguno) [inline docs/installation/server-config/avro-schema.rst]: Not just in CE, but in both COS and CE.

File: docs/installation/server-config/avro-schema.rst [hunk 6]
```diff
@@ -1,4 +1,96 @@
...
+
+- Read from Kafka topics with Avro-formatted data by using ``CREATE STREAM`` and ``CREATE TABLE`` statements.
+- Write Avro-formatted data into Kafka topics by using ``CREATE STREAM AS`` and ``CREATE TABLE AS`` statements.
```
Comment (miguno) [inline docs/installation/server-config/avro-schema.rst]: Better:

- Write Avro-formatted data into Kafka topics by using ``CREATE STREAM AS SELECT`` and ``CREATE TABLE AS SELECT`` statements.

And add:
- Create new, derived streams and tables from existing streams and tables with ``CREATE STREAM AS SELECT`` and ``CREATE TABLE AS SELECT`` statements.
- Convert easily between different data formats (e.g., convert a stream from Avro to JSON) with ``CREATE STREAM AS SELECT`` and ``CREATE TABLE AS SELECT`` statements.

**Comment by miguno** [comment]
The url http://staging-docs.confluent.io.s3-website-us-west-2.amazonaws.com/PR/778/4.1.x/ksql/docs/installation/server-config/avro-schema.html is wrong? It redirects to http://staging-docs.confluent.io.s3-website-us-west-2.amazonaws.com/current/, which atm shows 4.0.0 content?

File: docs/tutorials/examples.rst [hunk 7]
```diff
@@ -290,128 +290,10 @@ zipcode for each user:
...
-Avro schema for the new ``pageviews_avro`` stream, and it will also
-register the schema with Confluent Schema Registry.
-
 Running KSQL
 ------------
 
-KSQL supports various :ref:`modes of operation <modes-of-operation>`, including a standalone mode and a client-server mode.
+KSQL supports client mode and server mode.
 
 Additionally, you can also instruct KSQL to execute a single statement
```
Comment (miguno) [inline docs/tutorials/examples.rst]: I think this section needs updating because IIRC the syntax for `ksql http://localhost:8088 --exec "SELECT * FROM pageviews LIMIT 5;"` has changed.  @dguy should know more because he was already asked by @ybyzek about this.

File: docs/tutorials/examples.rst [hunk 8]
```diff
@@ -290,128 +290,10 @@ zipcode for each user:
...
-Note how you only need to set ``VALUE_FORMAT`` to Avro to achieve the
-data conversion. Also, KSQL will automatically generate an appropriate
-Avro schema for the new ``pageviews_avro`` stream, and it will also
-register the schema with Confluent Schema Registry.
-
 Running KSQL
 ------------
 
-KSQL supports various :ref:`modes of operation <modes-of-operation>`, including a standalone mode and a client-server mode.
+KSQL supports client mode and server mode.
```
Comment (miguno) [inline docs/tutorials/examples.rst]: I'd just remove this line.

File: docs/tutorials/examples.rst [hunk 9]
```diff
@@ -290,128 +290,10 @@ zipcode for each user:
...
-    CREATE STREAM pageviews_avro
-      WITH (VALUE_FORMAT = 'AVRO') AS
-      SELECT * FROM pageviews_json;
-
-Note how you only need to set ``VALUE_FORMAT`` to Avro to achieve the
-data conversion. Also, KSQL will automatically generate an appropriate
-Avro schema for the new ``pageviews_avro`` stream, and it will also
-register the schema with Confluent Schema Registry.
-
 Running KSQL
```
Comment (miguno) [inline docs/tutorials/examples.rst]: Rename to "Executing single KSQL statements from the command line"

File: docs/tutorials/examples.rst (hunk 7 above)
Comment (ybyzek) [reply docs/tutorials/examples.rst]: Unfortunately, `exec` option was removed.  Now you can do this

```bin/ksql http://localhost:8080 <<EOF
> run script 'ksql.commands;
> exit
> EOF```

Comment (miguno) [reply docs/tutorials/examples.rst]: Ok, here's a suggested for this paragraph with updated examples
----
Apart from using the KSQL CLI or launching KSQL servers with the `--queries-file` configuration, you can also execute KSQL statements from the command line.

Example 1:

```
$ ksql http://localhost:8088 <<EOF
> SHOW TOPICS;
> exit
> EOF
```
Example 2:

```
$ ksql http://localhost:8088 <<EOF
> RUN SCRIPT '/path/to/application.sql';
> exit
> EOF
```

File: docs/installation/server-config/avro-schema.rst (hunk 5 above)
Comment (joel-hamill) [reply docs/installation/server-config/avro-schema.rst]: changed to generic `|cp|`

File: docs/installation/server-config/avro-schema.rst (hunk 6 above)
Comment (joel-hamill) [reply docs/installation/server-config/avro-schema.rst]: added, with grammar edits

**Comment by joel-hamill** [comment]
@miguno i have addressed all feedback

File: docs/faq.rst [hunk 10]
```diff
@@ -78,9 +78,7 @@ KSQL currently supports formats:
 
 -  DELIMITED (e.g. comma-separated value)
 -  JSON
--  Avro (requires Confluent Schema Registry and setting ``ksql.schema.registry.url`` in the KSQL configuration file)
-
-*Support for Apache Avro is expected soon.*
+-  Avro (requires Confluent Schema Registry and :ref:`setting ``ksql.schema.registry.url`` in the KSQL configuration file <install_ksql-avro-schema>`)
```
Comment (miguno) [inline docs/faq.rst]: Nit: in the KSQL *server* configuration file

I want us to be extra clear on this because it might still be confusing what you can/should configure in the server vs. CLI configs.

File: docs/faq.rst [hunk 11]
```diff
@@ -149,9 +147,9 @@ flag. For more information, see :ref:`common-configs`.
 How do I use Avro data and integrate with Confluent Schema Registry?
 ====================================================================
 
-Configure the ``ksql.schema.registry.url`` to point to Schema Registry (see :ref:`common-configs`).
+Configure the ``ksql.schema.registry.url`` to point to Schema Registry (see :ref:`install_ksql-avro-schema`).
```
Comment (miguno) [inline docs/faq.rst]: Nit: Configure the ``ksql.schema.registry.url`` *in the KSQL server configuration* to point to...

File: docs/includes/ksql-includes.rst [hunk 12]
```diff
@@ -1,7 +1,7 @@
 .. Avro note
 
-.. note:: To use Avro, you must have Confluent Schema Registry enabled and set ``ksql.schema.registry.url`` in your KSQL configuration file.
-
+.. note:: To use Avro, you must have Confluent Schema Registry enabled and set :ref:```ksql.schema.registry.url`` in your
+          KSQL configuration file <install_ksql-avro-schema>`.
```
Comment (miguno) [inline docs/includes/ksql-includes.rst]: Nit: KSQL *server* configuration file

File: docs/installation/server-config/avro-schema.rst [hunk 13]
```diff
@@ -1,4 +1,100 @@
...
+
+-  Message keys in Avro format are not supported. Message keys in KSQL are always interpreted as STRING format, which means
+   KSQL will ignore Avro schemas that have been registered for message keys.
+-  Avro schemas with nested fields are not supported yet. This is because KSQL does not yet support nested columns.
+
+Configuring KSQL for Avro
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You must configure the API endpoint of |sr| by setting ``ksql.schema.registry.url`` (default: ``http://localhost:8081``)
+in the KSQL configuration file (``<path-to-confluent>/etc/ksql/ksql-server.properties``) that you use to start the KSQL
```
Comment (miguno) [inline docs/installation/server-config/avro-schema.rst]: Nit: KSQL *server* configuration file.  We can then remove "that you use to start the KSQL Server".

File: docs/installation/server-config/avro-schema.rst [hunk 14]
```diff
@@ -1,4 +1,100 @@
...
+    :local:
+
+Supported functionality
+^^^^^^^^^^^^^^^^^^^^^^^
+
+KSQL currently supports Avro data in the Kafka message values.
+
+The following functionality is not supported yet:
+
+-  Message keys in Avro format are not supported. Message keys in KSQL are always interpreted as STRING format, which means
```
Comment (ybyzek) [inline docs/installation/server-config/avro-schema.rst]: Are they ignored or is an error generated?

File: docs/installation/server-config/avro-schema.rst [hunk 15]
```diff
@@ -1,4 +1,100 @@
...
+   KSQL will ignore Avro schemas that have been registered for message keys.
+-  Avro schemas with nested fields are not supported yet. This is because KSQL does not yet support nested columns.
+
+Configuring KSQL for Avro
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You must configure the API endpoint of |sr| by setting ``ksql.schema.registry.url`` (default: ``http://localhost:8081``)
```
Comment (ybyzek) [inline docs/installation/server-config/avro-schema.rst]: Is it API endpoint or REST endpoint?

File: docs/installation/server-config/avro-schema.rst [hunk 16]
```diff
@@ -1,4 +1,100 @@
...
+in the KSQL configuration file (``<path-to-confluent>/etc/ksql/ksql-server.properties``) that you use to start the KSQL
+Server. For more information, see :ref:`install_ksql-server`.
+
+.. important:: Do not use the SET statement in the KSQL CLI to configure the registry endpoint.
+
+Using Avro in KSQL
+^^^^^^^^^^^^^^^^^^
+
+Before using Avro in KSQL, make sure that |sr| is up and running and that ``ksql.schema.registry.url`` is set correctly
+in the KSQL properties file. |sr| is :ref:`installed by default <quickstart>` with |cp|.
```
Comment (ybyzek) [inline docs/installation/server-config/avro-schema.rst]: Consider "installed" -> "included"

File: docs/installation/server-config/avro-schema.rst [hunk 17]
```diff
@@ -1,4 +1,100 @@
...
+
+Example: Create a new stream ``pageviews`` by reading from a Kafka topic with Avro-formatted messages.
+    .. code:: sql
+
+        CREATE STREAM pageviews
+          WITH (KAFKA_TOPIC='pageviews-avro-topic',
+                VALUE_FORMAT='AVRO');
+
+Example: Create a new table ``users`` by reading from a Kafka topic with Avro-formatted messages.
+    In this example you don’t need to define any columns or data types in the CREATE statement because KSQL automatically
```
Comment (ybyzek) [inline docs/installation/server-config/avro-schema.rst]: When you create a new stream by reading from a Kafka topic with Avro-formatted messages, you don't have to define columns or data types in the CREATE statement....if you are interacting via the CLI.  If you are putting these commands in a script that you then pass in via `queries-file`, then I think you do.  cc: @hjafarpour

File: docs/installation/server-config/avro-schema.rst (hunk 14 above)
Comment (joel-hamill) [reply docs/installation/server-config/avro-schema.rst]: @hjafarpour ^

File: docs/installation/server-config/avro-schema.rst (hunk 15 above)
Comment (joel-hamill) [reply docs/installation/server-config/avro-schema.rst]: @hjafarpour ^

File: docs/installation/server-config/avro-schema.rst (hunk 14 above)
Comment (hjafarpour) [reply docs/installation/server-config/avro-schema.rst]: The key schemas will be ignored and the key will be read using StringDeserializer.

File: docs/installation/server-config/avro-schema.rst [hunk 18]
```diff
@@ -1,4 +1,100 @@
...
+Supported functionality
+^^^^^^^^^^^^^^^^^^^^^^^
+
+KSQL currently supports Avro data in the Kafka message values.
+
+The following functionality is not supported yet:
+
+-  Message keys in Avro format are not supported. Message keys in KSQL are always interpreted as STRING format, which means
+   KSQL will ignore Avro schemas that have been registered for message keys.
+-  Avro schemas with nested fields are not supported yet. This is because KSQL does not yet support nested columns.
```
Comment (hjafarpour) [inline docs/installation/server-config/avro-schema.rst]: Maybe mention that this will be supported soon.

File: docs/installation/server-config/avro-schema.rst (hunk 15 above)
Comment (hjafarpour) [reply docs/installation/server-config/avro-schema.rst]: This is the REST endpoint for Schema Registry.

File: docs/installation/server-config/avro-schema.rst [hunk 19]
```diff
@@ -1,4 +1,100 @@
...
+in the KSQL server configuration file (``<path-to-confluent>/etc/ksql/ksql-server.properties``). For more information,
+see :ref:`install_ksql-server`.
+
+.. important:: Do not use the SET statement in the KSQL CLI to configure the registry endpoint.
+
+Using Avro in KSQL
+^^^^^^^^^^^^^^^^^^
+
+Before using Avro in KSQL, make sure that |sr| is up and running and that ``ksql.schema.registry.url`` is set correctly
+in the KSQL properties file. |sr| is :ref:`included by default <quickstart>` with |cp|.
```
Comment (hjafarpour) [inline docs/installation/server-config/avro-schema.rst]: Maybe mention that if `ksql.schema.registry.url` is not set we will use the default value, `http://localhost:8081`

File: docs/installation/server-config/avro-schema.rst [hunk 20]
```diff
@@ -1,4 +1,100 @@
...
+
+- Read from Kafka topics with Avro-formatted data by using ``CREATE STREAM`` and ``CREATE TABLE`` statements.
```
Comment (hjafarpour) [inline docs/installation/server-config/avro-schema.rst]: Maybe reword it to: 
`You can declare Stream or Table on Kafka topics with Avro-formatted data by using  CREATE STREAM and CREATE TABLE statements.`

File: docs/installation/server-config/avro-schema.rst [hunk 21]
```diff
@@ -1,4 +1,100 @@
...
+- Write Avro-formatted data into Kafka topics by using ``CREATE STREAM AS SELECT`` and ``CREATE TABLE AS SELECT`` statements.
```
Comment (hjafarpour) [inline docs/installation/server-config/avro-schema.rst]: Reword this to:
`You can read from and write into Avro-formatted data by using ...`

File: docs/installation/server-config/avro-schema.rst (hunk 17 above)
Comment (hjafarpour) [reply docs/installation/server-config/avro-schema.rst]: It should work in both cases. @ybyzek can you test it with `queries-file`?

File: docs/tutorials/examples.rst [hunk 22]
```diff
@@ -290,135 +290,50 @@ zipcode for each user:
...
-.. code:: shell
+        $ ksql http://localhost:8088 <<EOF
+        > RUN SCRIPT '/path/to/local/application.sql';
+        > exit
+        > EOF
 
-    $ ksql http://localhost:8088 --exec "SELECT * FROM pageviews LIMIT 5;"
\ No newline at end of file
+    .. note:: The ``RUN SCRIPT`` command only supports a subset of KSQL CLI commands, such as running persistent queries (CREATE
+              STREAM, CREATE TABLE), and setting configuration options (SET statement). It cannot show output for statements,
```
Comment (hjafarpour) [inline docs/tutorials/examples.rst]: Also add CSAS and CTAS to this.

File: docs/installation/server-config/avro-schema.rst (hunk 14 above)
Comment (joel-hamill) [reply docs/installation/server-config/avro-schema.rst]: so is there a doc change
 here?

Comment (hjafarpour) [reply docs/installation/server-config/avro-schema.rst]: This seems fine.

File: docs/tutorials/examples.rst (hunk 22 above)
Comment (hjafarpour) [reply docs/tutorials/examples.rst]: such as running persistent queries (CREATE STREAM AS SELECT, CREATE TABLE AS SELECT), DDL statements (CREATE STREAM, CREATE TABLE) and setting ...

Comment (miguno) [reply docs/tutorials/examples.rst]: And: it is not only hiding the output. It isn’t even executing those other statements.
